debian/changelog

```diff
@@ -1,3 +1,15 @@
+ola (0.11.0-1) UNRELEASED; urgency=medium
```

Comment:
The version parser in dpkg sorts ~ before anything, even an empty part. Therefore, according to dpkg,

    0~0 < 0~1 < 0

With that background, it makes sense to render this version number as `0.11.0~git-1`, so that anyone who installs a git snapshot doesn't have to wait for an "official" `0.11.0-2` (or above) before their system is upgraded.

Alternatively, a script could be run that uses the `dch` command from the `devscripts` package to generate a new changelog entry from CI configuration, which would use the output from `git describe` to generate a relevant version number; e.g., something like `0.11.0~git<git describe output>-1` could work. If this is combined with uploading the resulting package into one of the many package repository services out there, then this will allow people to install a git snapshot, get automatic upgrades, and then eventually upgrade to the "proper" released package, once it exists.

Reply:
> With that background, it makes sense to render this version number as 0.11.0~git-1, so that anyone who installs a git snapshot doesn't have to wait for an "official" 0.11.0-2 (or above) before their system is upgraded.

I really like this idea, though it is ultimately up to @peternewman.

> Alternatively, a script could be run that uses the `dch` command from the `devscripts` package to generate a new changelog entry from CI configuration, which would use the output from `git describe` to generate a relevant version number; e.g., something like `0.11.0~git<git describe output>-1` could work.

This is definitely more involved and could go in a later pull (manual versioning is probably fine for now). This mirrors the way that `setuptools_scm` works for Python/Pypi projects that I have used before and it is fantastic (you get unique build numbers even by commit hash...so useful!). This would play directly into dumping these builds into a nightly or dev build Debian repo!

Reply:
I'm going to work on this soon. I think of all of the pulls I have worked on recently this is the last one blocking me from using my own forked builds vs official builds in production.

Reply:
Yeah I love @yoe 's idea, even more so if we can somehow add it into bits of https://github.com/OpenLightingProject/ola/blob/master/common/base/Version.cpp and similar, so it shows up at the top of the log (and maybe on the webserver), given currently someone running master looks the same as someone running 0.10 branch. The challenge will be not forcing it to recompile each time, so it might just be a case of including the branch unless we give it some special flag for when its doing a package build.

I hadn't even considered pushing it to a repo, just rolling our own, is there one you'd recommend/suggest @yoe ?

Reply:
> The challenge will be not forcing it to recompile each time, so it might just be a case of including the branch unless we give it some special flag for when its doing a package build.

I have some web applications that check for the presence of `.git/FETCH_HEAD` at runtime to get the last commit hash (this doesn't require forking/executing the Git command during runtime). Though given that OLA is a project centered around Make/Autoconf, I think the best way would be to bake this into `Version.h` during the build. This way there is no dependency on Git during runtime.

Here is what I'm thinking:

1. When arbitrarily building, use the output of `git describe` or `.git/FETCH_HEAD` to generate a unique build identifier/version.
2. If the current directory is not a Git repo (missing `.git`), then fallback to the current method. This is useful if someone downloads a release source tarball instead of cloning and checking out. In this case they should still see the correct release number.
3. We add a flag to `./configure` that _forces_ the build to use the current versioning method so that when we build on a new tag/release we get a more official `0.11~git-1` build instead of `0.11~git-aabbcc-1` like we would want for nightlies and pull request builds. It's possible that `git describe` will also just work in this situation, but I need to read the docs on that.

Separately, we need to see what is up with `git describe` on this repo. I'm getting `0.10.2-3549-ge398f6d9b` on the `master` branch of my fork and I don't know what is causing that.

> I hadn't even considered pushing it to a repo, just rolling our own, is there one you'd recommend/suggest @yoe ?

I have seen a few repos that build in GitHub Actions and then push to the `gh-pages` branch and it gets automatically served by GitHub! It could push to this repo or we could have a separate repo for the builds. I would be really excited to have a nightly build APT repo as well as a release APT repo.

Other than that there are of course things like PPA, OBS, etc., though I am not familiar with the Debian-specific flavor (interestingly OBS can build most platforms).

Reply:
OK, part one is done: baking a build name into the Autoconf build. I left all of the existing versioning bit alone and simply added a new "build name" flag. I have it showing up on `./configure`, in `olad --version`, and all three web interfaces.

It uses `git describe` (this is all behavior from Git; I did not come up with this scheme except for the unknowns):

1. If Git is missing,the build name will be `unknown`.
2. If the folder is not a Git repo (we can't infer where it came from), then the build name will be `unknown-out-of-tree`. The existing versioning system still applies; this is just the build name.
3. If the Git repo is checked out to a tag or there are no new commits since the last tag, then the build name will simply be `0.10.9`. This means that when simply checking out a release it will reflect the untouched build in the build name.
4. If there have been any commits since the last tag, then Git describe will append `numberOfCommits-commitHash` to the end of the tag name. In this case the tag name is the most recent tag traced the current branch. So building in `master` would result in for example `0.10.9-9-gaabbccd` if there have been 9 commits since 0.10.9.
5. If there are changed but uncommited files in the repo, then Git will append `-dirty` to the build name.
6. Please don't name a tag `unknown` or `unknown-out-of-tree` :pleading_face: 

Examples:

- `0.10.9`: Checked out to the 0.10.9 tag
- `0.10.9-dirty`: Same as above but some files have changed and not been commited
- `0.10.9-2073-g0bf40e7`: Checked out to a branch that the last traceable tag as 0.10.9 and there have been 2073 commits since that
- `0.10.9-2073-g0bf40e7-dirty`: Same as above but some files have changed and not been commited
- `unknown-out-of-tree`: Folder is not a Git repository
- `unknown`: Git command is missing

I will now work on getting this into the Debian build!

Reply:
OK, this is working! The CI generates files like:
```
libola-dev_0.11.0~git-8482527-1_amd64.deb
libola1-dbgsym_0.11.0~git-8482527-1_amd64.deb
libola1_0.11.0~git-8482527-1_amd64.deb
ola-dbgsym_0.11.0~git-8482527-1_amd64.deb
ola-python_0.11.0~git-8482527-1_all.deb
ola-rdm-tests_0.11.0~git-8482527-1_all.deb
ola_0.11.0~git-8482527-1_amd64.buildinfo
ola_0.11.0~git-8482527-1_amd64.changes
ola_0.11.0~git-8482527-1_amd64.deb
```
I can also confirm that `debchange` indeed thinks this is a lower version than `0.11.0`.

It is only including the commit hash because the checkout action only checks out one commit at a time by default (it does not clone the tags to compare with). We can change this easily if we want the full/longer `git describe` output.

Reply:
UPDATE: OK, I enabled checking out all tags so we get builds like `0.11.0~git-0.10.9-2088-g69dcfdc-1`.

I realized the number of commits since last tag is useful because it will (usually) facilitate version increases between builds. The commit hash will never increment cleanly, but the number of commits always will due to Git's linearity. 

Example: `0.11.0~git-g1122334-1 < 0.11.0~git-g2233445-1` even if `0.11.0~git-g1122334-1` came later, but `0.11.0~git-0.10.9-124-g1122334-1 > 0.11.0~git-0.10.9-123-g2233445-1`.

Reply:
OK, I have added the version to the GitHub Actions build artifact names.

Reply:
> I hadn't even considered pushing it to a repo, just rolling our own, is there one you'd recommend/suggest @yoe ?

I tend to use [reprepro](https://packages.debian.org/reprepro), but setting that up in a CI system is involved, as it requires state be retained on the filesystem.

I have not that much experience with any of the repository-as-a-service thingies out there, so can't really help with that.